src/RestSharp/Request/RequestContent.cs

```diff
@@ -36,38 +36,66 @@ class RequestContent : IDisposable {
         _parameters = new RequestParameters(_request.Parameters.Union(_client.DefaultParameters));
     }
 
-    public HttpContent BuildContent() {
-        AddFiles();
-        var postParameters = _parameters.GetContentParameters(_request.Method).ToArray();
-        AddBody(postParameters.Length > 0);
-        AddPostParameters(postParameters);
+    public HttpContent BuildContent() 
+    {
+     if (_request.AlwaysMultipartFormData && _request.AlwaysSingleFileAsContent) 
+        throw new ArgumentException("Failed to put file as content because flag AlwaysMultipartFormData enabled");
+                
+        var postParameters       = _parameters.GetContentParameters(_request.Method).ToArray();
+        var postParametersExists = postParameters.Length > 0;
+        var bodyParametersExists = _request.TryGetBodyParameter(out var bodyParameter);
+        var filesExists          = _request.Files.Any();
+        
+        if (filesExists)
+            AddFiles(postParametersExists, bodyParametersExists);
+        
+        if (bodyParametersExists)
+            AddBody(postParametersExists, bodyParameter!);
+        
+        if (postParametersExists)
+            AddPostParameters(postParameters);
+     
         AddHeaders();
 
         return Content!;
     }
-
-    void AddFiles() {
-        if (!_request.HasFiles() && !_request.AlwaysMultipartFormData) return;
-
-        var mpContent = CreateMultipartFormDataContent();
-
-        foreach (var file in _request.Files) {
-            var stream = file.GetFile();
-            _streams.Add(stream);
-            var fileContent = new StreamContent(stream);
-
-            fileContent.Headers.ContentType = file.ContentType.AsMediaTypeHeaderValue;
-
-            var dispositionHeader = file.Options.DisableFilenameEncoding
-                ? ContentDispositionHeaderValue.Parse($"form-data; name=\"{file.Name}\"; filename=\"{file.FileName}\"")
-                : new ContentDispositionHeaderValue("form-data") { Name = $"\"{file.Name}\"", FileName = $"\"{file.FileName}\"" };
-            if (!file.Options.DisableFileNameStar) dispositionHeader.FileNameStar = file.FileName;
-            fileContent.Headers.ContentDisposition = dispositionHeader;
-
-            mpContent.Add(fileContent);
-        }
-
-        Content = mpContent;
+    
+     void AddFiles(bool postParametersExists, bool bodyParametersExists) 
+     {
+         // File uploading without multipart/form-data
+         if (_request.AlwaysSingleFileAsContent && _request.Files.Count == 1)
+         {
+             if (postParametersExists) throw new ArgumentException("Failed to put file as content because added post parameters");
```

Comment:
Just got an idea that we can create some sort of request validator in the request class itself, and it can be called each time anything gets changed there. Then, these rules will be in the request instead of checking it in many places.

Reply:
Agreed, but doing validation in the Execute method seems like a good solution